### PR TITLE
[Fix #4156] Keep max for Metrics in rubocop_todo

### LINKED ISSUE #4156: Unexpected behaviour when re-generating robocop_todo for Metric cops
When having a `.rubocop.yml` file with the following code:

``` Ruby
Metrics/LineLength:
  Max: 150
```
--------

## Actual behavior

if we add new files to the project where one of then has a line with more than 150 characters (let's say 200) and we want to re-generate the `rubocop_todo.yml` file, that's what we get:

``` Ruby
Metrics/LineLength:
  Max: 200
```

## Expected behavior

I expected to get:

``` Ruby
Metrics/LineLength:
  Max: 150
  Exclude:
     - 'file'
```

If not from that moment all in the rest of the files Rubocop won't complain of lines between 150 and 200 characters, what it did before. :confounded: 

## RuboCop version

```
0.47.1 (using Parser 2.4.0.0, running on ruby 2.2.5 x86_64-linux-gnu)
```

### PR DESCRIPTION
**Pull request by Ana06**
When generating the rubocop_todo.yml do not change the max for the metrics cops unless they exceed the exclude limit. :bowtie

Closes https://github.com/bbatsov/rubocop/issues/4156

The approach of having a hash for `:exclude_limit` may not be really elegant, but I do not come up with any other alternative that doesn't imply having track of the number of files with offenses and the limit of files with offenses in the cop class. The other good thing is that this approach could be reuse for other cops. **I have to fix the tests but I want to ensure that this approach is fine before doing it as it breaks quite a lot of tests.**

### CODE REVIEW

**Comment by jonas054** [comment]
The implementation with a hash seems OK to me. Looks like you didn't have to add too much code.

The functionality on the other hand, I'm not so sure about. Thinking only about RuboCop's own `.rubocop_todo.yml`, I think I prefer the current `Max` settings over `Exclude` settings. Should we add a command line option so people can choose?

**Comment by bbatsov** [comment]
Sounds like a reasonable idea to me, @jonas054.

File: lib/rubocop/cop/mixin/configurable_max.rb [hunk 1]
```diff
@@ -9,8 +9,10 @@ module ConfigurableMax
 
       def max=(value)
         cfg = config_to_allow_offenses
-        value = [cfg[max_parameter_name], value].max if cfg[max_parameter_name]
-        cfg[max_parameter_name] = value
+        current_max = cfg.dig(:exclude_limit, max_parameter_name)
```
Comment (rrosenblum) [inline lib/rubocop/cop/mixin/configurable_max.rb]: `dig` was introduced in Ruby 2.3. We still support down to Ruby 2.1 so you will have to use `cfg[:exclude_limit][max_parameter_name]`.

File: lib/rubocop/formatter/disabled_config_formatter.rb [hunk 2]
```diff
@@ -140,23 +140,28 @@ def default_config(cop_name)
       end
 
       def output_cop_config(output_buffer, cfg, cop_name)
+        offending_files = @files_with_offenses[cop_name].uniq.sort
+        # Only modify the maximum if the files are not excluded one by one.
+        if offending_files.count > @exclude_limit && cfg[:exclude_limit]
```
Comment (rrosenblum) [inline lib/rubocop/formatter/disabled_config_formatter.rb]: Use `size` instead of `count`. `count` should only be used when you have a block.

**Comment by rrosenblum** [comment]
> Should we add a command line option so people can choose?

I think a command line option or a setting on the `Metrics` cops would make sense. 

If you want `Metrics/LineLength` to be locked to a certain value, shouldn't it be in `.rubocop.yml` instead of `.rubocop_todo.yml`?

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

**Pull request closed without merging**